juju/juju
# [JUJU_968] Container provisioner worker count

Implement a container provisioner worker number model config for use with the container provisioner.  While investigating LP 1945813, I found that there were numerous lxd errors caused by the machine being unable to handle the number of containers to provision at once time. Not all machines can handle creating 16 containers in parallel well. Especially when deploy large bundles like the Openstack one to machines with less than 12G of memory.

Allow for the number of container provisioner task workers to be configured separately from the machine provisioner task workers.

Fix a bug where updating the model-config does not change the number of container provisioner workers.  Found while adding the above.

Improve debug messages, it's handy to know what machine exactly each worker is handling. Moved a few messages to Trace.

## QA steps

```sh
$ juju bootstrap localhost testme 
$ juju model-config logging-config="<root>=INFO;juju.worker.provisioner=TRACE"
$ juju model-config logging-config="<root>=INFO;juju.worker.provisioner=TRACE" -m controller

$ juju model-config num-container-provision-workers
4
$ juju model-config num-container-provision-workers=2
$ juju debug-log --include-module juju.worker.provisioner  # check log for:
machine-0: 20:01:46 INFO juju.worker.provisioner resizing provision worker pool size to 2

$ juju add-machine --series bionic
$ juju add-machine --series bionic lxd:1; juju add-machine --series bionic lxd:1; juju add-machine --series bionic lxd:1

$ juju debug-log --include-module juju.worker.provisioner # check log to see only 2 containers provisioned at once
machine-1: 20:04:29 DEBUG juju.worker.provisioner worker 1: processing task "start-instance 1/lxd/1"
machine-1: 20:04:29 DEBUG juju.worker.provisioner worker 0: processing task "start-instance 1/lxd/0"
machine-1: 20:04:53 INFO juju.worker.provisioner started machine 1/lxd/1 as instance juju-36d1b7-1-lxd-1 with hardware "", network config [], volumes [], volume attachments map[], subnets to zones [], lxd profiles []
machine-1: 20:04:53 DEBUG juju.worker.provisioner worker 1: processing task "start-instance 1/lxd/2"
machine-1: 20:04:53 INFO juju.worker.provisioner started machine 1/lxd/0 as instance juju-36d1b7-1-lxd-0 with hardware "", network config [], volumes [], volume attachments map[], subnets to zones [], lxd profiles []
machine-1: 20:05:07 INFO juju.worker.provisioner started machine 1/lxd/2 as instance juju-36d1b7-1-lxd-2 with hardware "", network config [], volumes [], volume attachments map[], subnets to zones [], lxd profiles []
```
